lib/src/shelf_run.dart

```diff
@@ -38,26 +40,46 @@ Future<ShelfRunContext> shelfRun(
 
   if (useHotReload) {
     withHotreload(() async {
-      final server = await _createServer(
+      try {
+        final server = await _createServer(
+          init,
+          defaultBindPort: defaultBindPort,
+          defaultBindAddress: defaultBindAddress,
+          defaultShared: defaultShared,
+          securityContext: securityContext,
+          onStarted: onStarted,
+        );
+        context._server = server;
+      }
+      on SocketException catch(e){
+        if(onStartFailed != null) {
+          onStartFailed.call(e);
+        }
+        else{
+          print(e);
+        }
+      }
+      return Future.value(context._server);
+    });
+  } else {
+    try {
+      await _createServer(
         init,
         defaultBindPort: defaultBindPort,
         defaultBindAddress: defaultBindAddress,
         defaultShared: defaultShared,
         securityContext: securityContext,
         onStarted: onStarted,
       );
-      context._server = server;
-      return server;
-    });
-  } else {
-    await _createServer(
-      init,
-      defaultBindPort: defaultBindPort,
-      defaultBindAddress: defaultBindAddress,
-      defaultShared: defaultShared,
-      securityContext: securityContext,
-      onStarted: onStarted,
-    );
+    }
+    on SocketException catch(e){
```

Comment:
the new ability (when e.g. an exception occurs while starting the server => custom action trigger) should reflect within a unit test (`test/`)